src/hotspot/share/oops/constantPool.hpp

```diff
@@ -914,88 +914,19 @@ class ConstantPool : public Metadata {
   const char* internal_name() const { return "{constant pool}"; }
 };
 
-class SymbolHashMapEntry : public CHeapObj<mtSymbol> {
- private:
-  SymbolHashMapEntry* _next;   // Next element in the linked list for this bucket
-  Symbol*             _symbol; // 1-st part of the mapping: symbol => value
-  unsigned int        _hash;   // 32-bit hash for item
-  u2                  _value;  // 2-nd part of the mapping: symbol => value
-
- public:
-  unsigned   int hash() const             { return _hash;   }
-  void       set_hash(unsigned int hash)  { _hash = hash;   }
-
-  SymbolHashMapEntry* next() const        { return _next;   }
-  void set_next(SymbolHashMapEntry* next) { _next = next;   }
-
-  Symbol*    symbol() const               { return _symbol; }
-  void       set_symbol(Symbol* sym)      { _symbol = sym;  }
-
-  u2         value() const                {  return _value; }
-  void       set_value(u2 value)          { _value = value; }
-
-  SymbolHashMapEntry(unsigned int hash, Symbol* symbol, u2 value)
-    : _next(NULL), _symbol(symbol), _hash(hash), _value(value) {}
-
-}; // End SymbolHashMapEntry class
-
-
-class SymbolHashMapBucket : public CHeapObj<mtSymbol> {
-
-private:
-  SymbolHashMapEntry*    _entry;
-
-public:
-  SymbolHashMapEntry* entry() const         {  return _entry; }
-  void set_entry(SymbolHashMapEntry* entry) { _entry = entry; }
-  void clear()                              { _entry = NULL;  }
-
-}; // End SymbolHashMapBucket class
-
-
 class SymbolHashMap: public CHeapObj<mtSymbol> {
-
- private:
-  // Default number of entries in the table
-  enum SymbolHashMap_Constants {
-    _Def_HashMap_Size = 256
-  };
-
-  int                   _table_size;
-  SymbolHashMapBucket*  _buckets;
-
-  void initialize_table(int table_size);
+  ResourceHashtable<const Symbol*, u2, 256, ResourceObj::C_HEAP, mtSymbol, Symbol::compute_hash> _table;
 
  public:
-
-  int table_size() const        { return _table_size; }
-
-  SymbolHashMap()               { initialize_table(_Def_HashMap_Size); }
-  SymbolHashMap(int table_size) { initialize_table(table_size); }
-
-  // hash P(31) from Kernighan & Ritchie
-  static unsigned int compute_hash(const char* str, int len) {
-    unsigned int hash = 0;
-    while (len-- > 0) {
-      hash = 31*hash + (unsigned) *str;
-      str++;
-    }
-    return hash;
+  void add_entry(const Symbol* sym, u2 value) {
+    bool created;
+    _table.put_if_absent(sym, value, &created);
```

Comment:
Pre-existing: Just an observation but it is very poor API naming to have `add_entry` when the semantics are actually `add_if_absent`.